### ISSUE TITLE
Restoring w/ Adam?

### ISSUE THREAD

**Issue by slerman12** [deepmind/sonnet#108]
I'm using `tf.train.Saver()` to save and restore a model. This works, except when training resumes, the accuracy of the model at first continues as it left off, but then suddenly drops completely as if starting from scratch. I checked everything in my code and I don't know what else could be responsible for it. I am using Adam and I wonder if those variables are not being restored because of some nuance of Sonnet. I can't figure out why training accuracy drops after a few batches with the restored model.

**Comment by malcolmreynolds** [comment]
Can you verify that the Adam moving average variables are actually in the checkpoint? If you create a saver for the Sonnet module before you've instantiated the optimizer, this will only make save / restore nodes for the weights, not the associated optimization variables. You can check if they are getting restored by printing out some values of both the weights and optimization variables before and after the `saver.restore()` call.

**Comment by slerman12** [comment]
It did not print all of the Sonnet variables or the non-Sonnet LSTM variables.

**Comment by malcolmreynolds** [comment]
Is the declaration of the tf.Saver after the call to optimizer.minimize(loss)? That's the point when the auxiliary variables get created, I would think.

If you print out `saver._var_list` after it's been created, you can see whether the Adam variables are in the checkpoint or not. For what it's worth I just did a quick test with a sonnet module hooked to a toy loss, then made an AdamOptimizer and called minimize, then made a Saver. All the auxiliary Adam variables were definitely in `saver._var_list`, and were saving / restoring correctly for me.

**Comment by slerman12** [comment]
Yes, the saver is also after the train op. Looks like all of the variables are restored. I'm not sure why the training resumes properly but then suddenly goes down as if starting from scratch. Then it comes up again once it trains again.

My restoration code is:

```
with tf.Session() as sess:
    sess.run(init)

    directory_name = "directory"
    path = os.getcwd()
    log_directory = path + "/" + directory_name + "/"

    if not os.path.exists(log_directory + "Saved/"):
        os.makedirs(log_directory + "Saved/")
    if tf.train.checkpoint_exists(log_directory + "Saved/model"):
        saver.restore(sess, log_directory + "Saved/model")
```

Then I do normal batch training. Once every number of batches, I save with` saver.save(sess, log_directory + "Saved/model")`.

Seems the problem is not related to Sonnet, but I'm not sure what else it could be.

**Comment by slerman12** [comment]
Changing:

`# logits = snt.nets.mlp.MLP([256, 256, 256, 256, vocab_size])(x)`

to

`logits = snt.nets.mlp.MLP([vocab_size])(x)`

fixed the problem for me. I don't know if it's just a coincidence or if the stacked MLP layers somehow interfere with regular saving.

**Comment by slerman12** [comment]
Aha! My data was not being shuffled upon restore and the model would begin to overfit to highly correlated batches (more MLP layers allowed it to overfit more) - I should have realized. Thank you for your help.
